Simplify GPUExternalTexture WebCodecsVideoFrame implementation

mwyrzykowski opened the pull request:
#### f9d7fea52f903441640365484bf48dc5d50d82ff
<pre>
Simplify GPUExternalTexture WebCodecsVideoFrame implementation
<a href="https://bugs.webkit.org/show_bug.cgi?id=257724">https://bugs.webkit.org/show_bug.cgi?id=257724</a>
&lt;radar://110274148&gt;

Reviewed by Kimmo Kinnunen.

Address feedback from #14055

Specifically in this patch we no longer pass the identifier + version manually
from the WebProcess to GPUProcess, instead SharedFrameVideoReader/Writer are used instead.

Also some stylistic feedback addressed.

Testing covered by GPUExternalTexture CTS tests.

* Source/WebCore/Modules/WebGPU/GPUExternalTextureDescriptor.h:
(WebCore::GPUExternalTextureDescriptor::mediaIdentifierForSource):
(WebCore::GPUExternalTextureDescriptor::convertToBacking const):

* Source/WebCore/Modules/WebGPU/InternalAPI/WebGPUExternalTextureDescriptor.h:
* Source/WebCore/platform/VideoFrame.h:
(WebCore::VideoFrame::pixelBuffer const):
(WebCore::VideoFrame::resourceIdentifier const): Deleted.

* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.cpp:
(WebKit::RemoteDevice::importExternalTextureFromVideoFrame):
(WebKit::RemoteDevice::importExternalTexture): Deleted.
(WebKit::RemoteDevice::importExternalTextureFromPixelBuffer): Deleted.

* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.messages.in:
* Source/WebKit/Scripts/webkit/messages.py:
(types_that_must_be_moved):
(headers_for_type):

* Source/WebKit/Shared/WebGPU/WebGPUExternalTextureDescriptor.h:
* Source/WebKit/Shared/WebGPU/WebGPUExternalTextureDescriptor.serialization.in:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteDeviceProxy.cpp:
(WebKit::WebGPU::RemoteDeviceProxy::importExternalTexture):

* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteDeviceProxy.h:
* Source/WebKit/WebProcess/GPU/media/RemoteVideoFrameProxy.cpp:
(WebKit::RemoteVideoFrameProxy::pixelBuffer const):
(WebKit::RemoteVideoFrameProxy::resourceIdentifier const): Deleted.

 Source/WebKit/WebProcess/GPU/media/RemoteVideoFrameProxy.h:
* Source/WebKit/WebProcess/GPU/webrtc/MediaRecorderPrivate.cpp:
(WebKit::MediaRecorderPrivate::videoFrameAvailable):

Canonical link: <a href="https://commits.webkit.org/268232@main">https://commits.webkit.org/268232@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/381123a61d91acd6843f3fa52589b9d12af92db9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/19038 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/19381 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/19989 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/20903 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/17797 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/19225 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/22674 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/19550 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/19553 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/19261 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/19365 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/16566 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/21786 "Built successfully") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/19210 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/16574 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/17351 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/23747 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/17614 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/17522 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/21694 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/18095 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/15373 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/17183 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4547 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/21541 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/17930 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/PAL/pal/graphics/WebGPU/WebGPUExternalTextureDescriptor.h:
```diff
@@ -33,24 +33,20 @@
 
-namespace PAL::WebGPU {
+namespace WebCore {
+class VideoFrame;
```
We are in PAL, so we cannot really refer to VideoFrame.
We could probably move VideoFrame to PAL.

mwyrzykowski replied:
Agreed this is probably a layering violation which only works as this is a forward declaration. I will see how much effort it would be to move VideoFrame to PAL.

mwyrzykowski replied:
@youennf I updated the PR with creating a PAL::VideoFrame and added some static downcasts. I used `static_cast` as all`PAL::VideoFrame` instances are `WebCore::` ones to avoid any performance loss with `dynamicDowncast`. If you think there are any issues with this idea please let me know.

I initially attempted to move all of `VideoFrame` to PAL but this is a very large change as it pulls in several WebCore types. I think it is far too risky given where we are in the software development cycle.

youennf commented on Source/WebCore/PAL/pal/graphics/WebGPU/WebGPUExternalTextureDescriptor.h:
```diff
@@ -33,24 +33,20 @@
-    VideoSourceIdentifier mediaIdentifier;
+    uint64_t mediaIdentifier;
     PredefinedColorSpace colorSpace { PredefinedColorSpace::SRGB };
+    WebCore::VideoFrame* sharedFrame { nullptr };
```
Raw pointers like that are a bit frightening.
Having RefPtr<> would be better.

mwyrzykowski commented on Source/WebKit/GPUProcess/graphics/WebGPU/RemoteGPU.h:
```diff
@@ -115,12 +109,11 @@ class RemoteGPU final : public IPC::StreamMessageReceiver {
     void createCompositorIntegration(WebGPUIdentifier);
 
-    WeakPtr<GPUConnectionToWebProcess> m_gpuConnectionToWebProcess;
+    ThreadSafeWeakPtr<GPUConnectionToWebProcess> m_gpuConnectionToWebProcess;
```
@kkinnunen-apple is this change reasonable? `GPUConnectionToWebProcess` is created on the main thread but I want to access it from `RemoteDevice` on the WebGPU work queue

kkinnunen-apple replied:
No, it's quite incorrect.
```
ThreadSafeWeakPtr<GPUConnectionToWebProcess> m_gpuConnectionToWebProcess;
```

This means that when GPUConnectionToWebProcess is destroyed, m_gpuConnectionToWebProcess is nullptr.
However, you don't need this information, as GPUConnectionToWebProcess is the one that stops all WebGPU work before GPUConnectionToWebProcess is destroyed.

m_gpuConnectionToWebProcess in general is pointless variable, as you cannot do anything with it in the WebGPU work queue. The contents of GPUConnectionToWebProcess is not thread safe.

mwyrzykowski replied:
Sounds good I will revert these changes for `ThreadSafeWeakPtr`

mwyrzykowski replied:
Is the `m_gpuConnectionToWebProcess` really pointless however? I can still call `.videoFrameObjectHeap` which should be thread safe as `VideoFrameObjectHeap` is thread safe right?

And calling `gpuConnectionToWebProcess.performWithMediaPlayerOnMainThread(...` is thread safe too right? As `GPUConnectionToWebProcess::performWithMediaPlayerOnMainThread` calls `callOnMainRunLoopAndWait`. `callOnMainRunLoopAndWait` appears thread safe to me.

Alternatively I could pass the `VideoFrameObjectHeap` and `RemoteMediaPlayerManagerProxy` instance variables from `GPUConnectionToWebProcess` down to `RemoteDevice` to achieve the same thing. Would that be preferred?

kkinnunen-apple commented on Source/WebKit/WebProcess/GPU/webrtc/MediaRecorderPrivate.cpp:
```diff
@@ -89,7 +89,7 @@ MediaRecorderPrivate::~MediaRecorderPrivate()
 }
 
-void MediaRecorderPrivate::videoFrameAvailable(VideoFrame& videoFrame, VideoFrameTimeMetadata)
+void MediaRecorderPrivate::videoFrameAvailable(WebCore::VideoFrame& videoFrame, VideoFrameTimeMetadata)
```
Not needed, WebCore is already in scope.

mwyrzykowski replied:
I had to add this because of `using namespace PAL` at the top of the file results in a name clash as there is now `PAL::VideoFrame` and `WebCore::VideoFrame` so the compiler can not deduce the correct type without specifying `WebCore::` here

kkinnunen-apple commented on Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
```diff
@@ -654,22 +654,7 @@ void GPUConnectionToWebProcess::createRemoteGPU(WebGPUIdentifier identifier, Ren
-                    performWithMediaPlayerOnMainThread(i, [&callback](auto& player) {
-                        callback(player.videoFrameForCurrentTime());
-                    });
-                }, [&] (WebKit::RemoteVideoFrameReference i) {
```
So this function is quite incorrect for various reasons.

There's two use-cases that are distinct:
1. Getting content from "MediaPlayer"
2. Getting contetnt from "WebCodecsVideoFrame"

This function conflates the two.

Keep your work distinct, and only concentrate on 2.

For your current patch to improve "the various reasons":
The idea is:
1. WebCodecsVideoFrame video frames are stored in video frame object heap. VideoFrameObjectHeap is thread-safe, it is intended for cross work queue sharing. 
2. WebGPU must fetch the video frame from the object heap, based on the video frame read reference.


To be done:
1. Remove the variant thingy from this
2. Pass Ref<VideoFrameObjectHeap> to WebGPU objects 
3. Read from the video frame object heap whenever you have a video frame read reference.

Communicate the video frame read references from the Web process, do not do it like this thing here, where the read reference is constructed on-the-fly in GPUP based on the identifier.

kkinnunen-apple replied:
>Communicate the video frame read references from the Web process, do not do it like this thing here, where the read reference is constructed on-the-fly in GPUP based on the identifier.

This might be a bit hard to do currently, as the code apparently has been going into different direction. It needs to probably be fixed at some point to go to the original direction.

mwyrzykowski replied:
@kkinnunen-apple doesn’t SharedVideoFrameReader already use the RemoteVideoFrameObjecyHeap? I pass it to SharedVideoFrameReader‘s constructor in RemoteDevice.cpp with this patch

kkinnunen-apple commented on Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
```diff
@@ -1590,7 +1590,7 @@ ExceptionOr<void> CanvasRenderingContext2DBase::drawImage(WebCodecsVideoFrame& f
 
     // FIXME: Add support for srcRect
-    context->paintVideoFrame(*internalFrame, dstRect, frame.shoudlDiscardAlpha());
+    context->paintVideoFrame(static_cast<VideoFrame&>(*internalFrame), dstRect, frame.shoudlDiscardAlpha());
```
With a cursory look,  don't see a reason for this static cast after you revert the internalFrame() return value..

kkinnunen-apple commented on Source/WebCore/PAL/pal/platform/VideoFrame.h:
```diff
@@ -0,0 +1,41 @@
+
+namespace PAL {
+
+class VideoFrame : public ThreadSafeRefCounted<VideoFrame> {
```
It's a sorry state of affairs if we have to arrive to this redundant class.
The  media class hierarchy is already needlessly complicated. Based on thisI'll ask you to humour me and split the patch a bit, so you see the correct solution and add minimal clutter.

kkinnunen-apple commented on Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.messages.in:
```diff
@@ -47,7 +46,7 @@ messages -> RemoteDevice NotRefCounted Stream {
     void SetSharedVideoFrameSemaphore(IPC::Semaphore semaphore) NotStreamEncodable
     void SetSharedVideoFrameMemory(WebKit::SharedMemory::Handle storageHandle) NotStreamEncodable
-    void ImportExternalTextureFromPixelBuffer(WebKit::WebGPU::ExternalTextureDescriptor descriptor, std::optional<WebKit::SharedVideoFrame::Buffer> sharedBuffer, WebKit::WebGPUIdentifier identifier) NotStreamEncodable
+    void ImportExternalTextureFromVideoFrame(WebKit::WebGPU::ExternalTextureDescriptor descriptor, std::optional<WebKit::SharedVideoFrame> sharedFrame, WebKit::WebGPUIdentifier identifier) NotStreamEncodable
```
Here you have two things defining the frame:
 - descriptor
 - sharedFrame

Contrast:
 - Why is the information of sharedFrame *not* inside descriptor like other defining aspects of the message?
 - Why is the information of the descriptor *not* a parameter, like other defining aspects of the message, rather stored in a descriptor?


See below at WebKit::WebGPU::ExternalTextureDescriptor

mwyrzykowski replied:
I think this is yet to be addressed, I will take another look now

mwyrzykowski replied:
`sharedFrame` is not part of `ExternalTextureDescriptor` because its definition as an RValue in SharedVideoFrame.serialization.in appears to prevent it from being used in a struct with StreamConnectionEncoder.

But maybe I don't even need the descriptor and can just pass parameters. I will try that.

mwyrzykowski replied:
Passing all the information in the descriptor via parameters is somewhat fragile since this descriptor derives from a base type and any change there would require updated parameters.

Maybe @achristensen07 knows if we can use the auto-generated serialization for structs containing members of types like `SharedVideoFrame` which use the Rvalue specifier like so?
```
[RValue] struct WebKit::SharedVideoFrame
```

mwyrzykowski replied:
I forgot to add the containing type to `types_that_must_be_moved`, with that change it seems to work correctly now

kkinnunen-apple commented on Source/WebKit/Shared/WebGPU/WebGPUExternalTextureDescriptor.h:
```diff
@@ -36,7 +36,7 @@
 
 struct ExternalTextureDescriptor : public ObjectDescriptorBase {
-    PAL::WebGPU::VideoSourceIdentifier mediaIdentifier;
+    uint64_t mediaIdentifier;
```
So here is a definition of the external texture: obtain the texture from identifier.
.. except, if the message had sharedFrame, in that case obtain it from the shared frame.

This is wildly incoherent.

Note, your corresponding "ideal world model" is :

GPUExternalTextureDescriptor
* GPUVideoSource source; // should be renamed to GPUExternalTextureSource
* GPUPredefinedColorSpace colorSpace

So you should have corresponding:

WebKit::WebGPU::ExternalTextureDescriptor
* WebKit::WebGPU::ExternalTextureSource source
* PAL::WebGPU::PredefinedColorSpace colorSpace

using WebKit::WebGPU::ExternalTextureSource = std::variant<MediaPlayerIdentifier, WebKit::SharedVideoFrame>;

kkinnunen-apple commented on Source/WebCore/Modules/WebGPU/GPUExternalTextureDescriptor.h:
```diff
@@ -45,43 +45,33 @@ using GPUVideoSource = RefPtr<HTMLVideoElement>;
-                if (internalFrame->isRemoteProxy())
-                    return PAL::WebGPU::WebCodecsVideoFrameIdentifier { internalFrame->resourceIdentifier() };
+        , [](const RefPtr<WebCodecsVideoFrame> videoFrame) -> std::pair<uint64_t, RefPtr<PAL::VideoFrame>> {
+            if (auto internalFrame = videoFrame->internalFrame())
```
This deals with "remote" logic.
The remote logic is not part of WebCore, WebCore doesn't understand that.
This should be: 
- invoked polymorphically with two implementations: one in webcore (in-process), one in webkit (gpup)
- outcome should be PAL::WebGPU::ExternalTextureDescriptor contents that makes sense to be used polymorphically by the receiver of the object

HOWEVER:: due to the design mistake of WebGPU using PAL, this cannot be such, as PAL doesn't have anything to build the polymorphism with. We cannot pass WebCore types polymorphically in non-type-punning manner.

You are end up smuggling the polymorphic data with upcasting.
For this purpose, you should have some sort of union with void* to avoid confusions and contraptions.

kkinnunen-apple commented on Source/WebCore/Modules/webcodecs/WebCodecsVideoFrameData.h:
```diff
@@ -27,16 +27,16 @@
     size_t memoryCost() const { return 4 * codedWidth * codedHeight; }
 
-    RefPtr<VideoFrame> internalFrame;
+    RefPtr<PAL::VideoFrame> internalFrame;
```
This shouldn't be needed. There isn't any PAL::VideoFrame instances that aren't WebCore::VideoFrame instances.

kkinnunen-apple commented on Source/WebCore/platform/VideoEncoder.h:
```diff
@@ -55,7 +56,7 @@ class VideoEncoder {
     };
     struct RawFrame {
-        Ref<VideoFrame> frame;
+        Ref<PAL::VideoFrame> frame;
```
This shouldn't be needed. There isn't any PAL::VideoFrame instances that aren't WebCore::VideoFrame instances.

kkinnunen-apple commented on Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.cpp:
```diff
@@ -153,23 +159,17 @@ void RemoteDevice::createSampler(const WebGPU::SamplerDescriptor& descriptor, We
 }
 
-static void populateConvertedDescriptor(auto mediaIdentifier, auto& convertedDescriptor, auto& performWithMediaPlayerOnMainThread)
+static void populateConvertedDescriptor(auto mediaIdentifier, auto& convertedDescriptor, auto& gpuConnectionToWebProcess)
```
To simplify and really crystalize the hard part of this patch, could you submit the easy parts in isolation?
E.g. submit a simple patch of replacing m_performWithMediaPlayerOnMainThread with the m_gpuConnectionToWebProcess


Note also: freestanding function populateConvertedDescriptor is pointless function, as it is calling one method. It is called inside ifdef block but it itself also contains the same ifdef with an empty implementation.  Currently this is just COCOA logic all the way.

kkinnunen-apple reviewed (commented):
The  media class hierarchy is already needlessly complicated. Based on thisI'll ask you to humour me and split the patch a bit, so you see the correct solution and add minimal clutter.

The hard part of this patch is:
- WebGPU is based on PAL
- PAL does not have WebCore types
- PAL does not have the primitives to implement polymorphism that WebKit remote implementation needs due to the above
- You are introducing a refcounted type PAL::VideoFrame, purpose of which is to only 1) hold refcount when the ref is stored in a descriptor 2) be downcasted manually 

To simplify and really crystalize the hard part of this patch, could you submit the easy parts in isolation?

kkinnunen-apple commented on Source/WebCore/PAL/pal/graphics/WebGPU/WebGPUExternalTextureDescriptor.h:
```diff
@@ -35,22 +36,14 @@ typedef struct __CVBuffer* CVPixelBufferRef;
-
 struct ExternalTextureDescriptor : public ObjectDescriptorBase {
-    VideoSourceIdentifier mediaIdentifier;
+    uint64_t mediaIdentifier;
```
I think here it should really help crystallizing the problem for your thought when you would define this, too as
```
using ExternalTextureSource = std::variant</* what is this??? */>;

struct ExternalTextureDescriptor : public ObjectDescriptorBase {
    PredefinedColorSpace colorSpace { PredefinedColorSpace::SRGB };
    ExternalTextureSource source;
};
```

mwyrzykowski commented on Source/WebCore/Modules/WebGPU/InternalAPI/WebGPUExternalTextureDescriptor.h:
```diff
@@ -35,22 +36,14 @@ typedef struct __CVBuffer* CVPixelBufferRef;
-
 struct ExternalTextureDescriptor : public ObjectDescriptorBase {
-    VideoSourceIdentifier mediaIdentifier;
+    uint64_t mediaIdentifier;
```
Now that what was PAL is moved to WebCore, we should be able to use WebCore::MediaIdentifier here and at related places

kkinnunen-apple reviewed (approved):
Looks good

mwyrzykowski committed:
Simplify GPUExternalTexture WebCodecsVideoFrame implementation
https://bugs.webkit.org/show_bug.cgi?id=257724
<radar://110274148>

Reviewed by Kimmo Kinnunen.

Address feedback from #14055

Specifically in this patch we no longer pass the identifier + version manually
from the WebProcess to GPUProcess, instead SharedFrameVideoReader/Writer are used instead.

Also some stylistic feedback addressed.

Testing covered by GPUExternalTexture CTS tests.

* Source/WebCore/Modules/WebGPU/GPUExternalTextureDescriptor.h:
(WebCore::GPUExternalTextureDescriptor::mediaIdentifierForSource):
(WebCore::GPUExternalTextureDescriptor::convertToBacking const):

* Source/WebCore/Modules/WebGPU/InternalAPI/WebGPUExternalTextureDescriptor.h:
* Source/WebCore/platform/VideoFrame.h:
(WebCore::VideoFrame::pixelBuffer const):
(WebCore::VideoFrame::resourceIdentifier const): Deleted.

* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.cpp:
(WebKit::RemoteDevice::importExternalTextureFromVideoFrame):
(WebKit::RemoteDevice::importExternalTexture): Deleted.
(WebKit::RemoteDevice::importExternalTextureFromPixelBuffer): Deleted.

* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.messages.in:
* Source/WebKit/Scripts/webkit/messages.py:
(types_that_must_be_moved):
(headers_for_type):

* Source/WebKit/Shared/WebGPU/WebGPUExternalTextureDescriptor.h:
* Source/WebKit/Shared/WebGPU/WebGPUExternalTextureDescriptor.serialization.in:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteDeviceProxy.cpp:
(WebKit::WebGPU::RemoteDeviceProxy::importExternalTexture):

* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteDeviceProxy.h:
* Source/WebKit/WebProcess/GPU/media/RemoteVideoFrameProxy.cpp:
(WebKit::RemoteVideoFrameProxy::pixelBuffer const):
(WebKit::RemoteVideoFrameProxy::resourceIdentifier const): Deleted.

 Source/WebKit/WebProcess/GPU/media/RemoteVideoFrameProxy.h:
* Source/WebKit/WebProcess/GPU/webrtc/MediaRecorderPrivate.cpp:
(WebKit::MediaRecorderPrivate::videoFrameAvailable):

Canonical link: https://commits.webkit.org/268232@main
